nservicebus/messaging/sharing-contracts.md

```diff
@@ -0,0 +1,54 @@
+---
+title: Defining message contracts
+summary: How to define and share message contracts between different endpoints.
+component: Core
+reviewed: 2022-02-02
+related:
+ - nservicebus/messaging/conventionsmessages-events-commands
+ - nservicebus/messaging/unobtrusive-mode
+---
+
+TODO: add menu entry
+
+It's recommended to use dedicated assemblies for message contracts. By keeping message contracts in a separate assembly, the amount of information and dependencies shared between services is minimized. It is recommended to have a separate message assembly for every service. When doing so, a service can [evolve its contracts](/nservicebus/messaging/evolving-contracts.md) without impacting other services in the system. Every message contract should be declared in the contracts assembly of the service owning that message contract.
+
+## Sharing contracts
+
+Sender and receiver of a message need to use the same message contract. Message contracts can be shared in various ways:
+
+* When all endpoints are located in a single solution, the message contract assemblies can be directly referenced as a project dependency.
+* When endpoints are split into multiple solutions, message contracts can be shared as NuGet packages. Follow the [NuGet settings](TODO) to create a NuGet package when building the contracts assembly. The NuGet package can be deployed to a local folders, private or public NuGet server. Use a [custom nuget config file](TODO) to configure NuGet package sources.
+* It's also possible to share messages as C# source files without packaging them into an assembly.
+
+### NServiceBus.Core dependency
+
+when using the [marker interfaces](TODO) to define messages, the message contract project will have a dependency on the NServiceBus package. This can limit the ability to share new versions of message contract assemblies with older endpoints, as this might produce version conflicts when trying to update the message contracts on endpoints targeting older major versions of NServiceBus. In order to share the latest contract assembly version with all endpoints, consider these options:
+
+* Ensure that the message contract assembly always references the oldest used NServiceBus major version. Due to [Nuget dependency resolution](https://docs.microsoft.com/en-us/nuget/concepts/dependency-resolution), endpoints on newer major versions of NServiceBus can reference assemblies that target an older version of NServiceBus, but not vice versa.
+* Switch to [unobstrusive mode](/nservicebus/messaging/unobtrusive-mode.md). Unobtrusive mode allows message contracts that have no dependency on the NServiceBus package, making it very easy to share message contracts across different versions and platforms.
```

Comment:
Maybe we should make this guidance stronger and not just a bullet point in a section on the marker interfaces. Like come down more strongly saying "We want you to use unobtrusive mode, it's better"